dask/bytes/core.py

```diff
@@ -1,22 +1,104 @@
 from __future__ import print_function, division, absolute_import
 
 import io
+import os
 
 from toolz import merge
+from warnings import warn
 
 from .compression import seekable_files, files as compress_files
 from .utils import SeekableFile
-from ..compatibility import PY2
+from ..compatibility import PY2, unicode
 from ..delayed import delayed
-from ..utils import infer_storage_options, system_encoding
+from ..utils import (infer_storage_options, system_encoding,
+                     build_name_function, infer_compression)
 
 delayed = delayed(pure=True)
 
 # Global registration dictionaries for backend storage functions
 # See docstrings to functions below for more information
 _read_bytes = dict()
+_write_bytes = dict()
 _open_files = dict()
 _open_text_files = dict()
+_write_files = dict()
+
+
+def write_bytes(values, urlpath, name_function=None, compression=None,
+                lazy=False, **kwargs):
+    """For a list of values which evaluate to byte, produce delayed values
+    which, when executed, result in writing to files.
+
+    The path maybe a concrete directory, in which case it is interpreted
+    as a directory, or a template for numbered output.
+
+    The path may be preceded by a protocol, like ``s3://`` or ``hdfs://`` if
+    those libraries are installed.
+
+    Parameters
+    ----------
+    values: list of Values or dask collection
+        the data to be written
+    urlpath: string
+        Absolute or relative filepaths, URLs (may include protocols like
+        ``s3://``); may be globstring (include `*`).
+    name_function: function or None
+        If using a globstring, this provides the conversion from part number
+        to test to replace `*` with.
+    compression: string or None
+        String like 'gzip' or 'xz'.  Must support efficient random access.
+    **kwargs: dict
+        Extra options that make sense to a particular storage connection, e.g.
+        host, port, username, password, etc.
+
+    Examples
+    --------
+    >>> values = write_bytes(vals, 's3://bucket/part-*.csv')  # doctest: +SKIP
+
+    Returns
+    -------
+    list of ``dask.Delayed`` objects
+    """
+    if hasattr(values, 'to_delayed'):
+        # bag, array or dataframe collection
+        # TODO: better as an isinstance ?
+        values = list(values.to_delayed())
```

Comment:
I'm somewhat in favor of only accepting a list of delayed values and forcing the collections to think about this explicitly.  It reduces magic a bit which, for an internal function like this, might be nice.  Also, `dask.array.to_delayed` operates somewhat differently, returning a numpy array of delayed objects rather than a list.